### PR TITLE
Make machine tests nondestructive

### PR DESCRIPTION
**Pull request by marusak**
It needs https://github.com/cockpit-project/cockpit/pull/13696 (the cleaning of `prepareStorageDeviceOnISCSI`).

### CODE REVIEW

File: test/verify/machineslib.py [hunk 1]
```diff
@@ -1810,7 +1816,10 @@ def installWithErrorTest(dialog):
             self.machine.execute(" && ".join(cmds))
 
             # Add an extra network interface that should appear in the PXE source dropdown
-            iface = self.add_iface(activate=False)
+            self.machine.upload(["verify/files/veth.sh"], "/tmp/")
+            veth = self.machine.spawn("/tmp/veth.sh", "veth")
+            self.addCleanup(self.machine.execute, "kill {0} || true".format(veth))
+            iface = "eth42"
```
Comment (martinpitt) [inline test/verify/machineslib.py]: This is fine for initial testing of course, but for the final version this needs to move to testlib.py, similar to `add_disk()`, and not hardcode the name. It also doesn't need to be a spawned process, as it does nothing. It should just addCleanup the teardown of the interface name.

File: test/verify/check-machines-dbus [hunk 2]
```diff
@@ -1849,18 +1849,25 @@ class TestMachinesDBus(machineslib.TestMachines):
...
+        dir_pool = os.path.join(self.tmp_storage, "dir_pool")
+        m.execute("mkdir --mode 777 {0}".format(dir_pool))
+        m.execute("virsh pool-define-as dir-pool --type dir --target {0} && virsh pool-start dir-pool".format(dir_pool))
 
         # Prepare net-fs pool
-        m.execute("mkdir /mnt/nfs-pool && mkdir /mnt/exports && echo '/mnt/exports 127.0.0.1/24(rw,sync,no_root_squash,no_subtree_check,fsid=0)' > /etc/exports")
+        nfs_pool = os.path.join(self.tmp_storage, "nfs_pool")
+        mnt_exports = os.path.join(self.tmp_storage, "mnt_exports")
+        m.execute("mv /etc/exports /etc/exports.backup")
+        self.addCleanup(m.execute, "mv /etc/exports.backup /etc/exports")
```
Comment (martinpitt) [inline test/verify/check-machines-dbus]: We do this often enough that it should go into `TestMachines.setUp()` IMHO.

**Comment by martinpitt** [comment]
@marusak: #13701 landed, so this should be unblocked now.

**Comment by marusak** [comment]
If testCreate fails, then all subsequent runs fails as well as it fails to use a new certificate.

**Comment by marusak** [comment]
It creates certificate with:
```
$ openssl req -x509 -newkey rsa:4096 -keyout key.pem -out cert.pem -days 365 -subj '/CN=localhost' -nodes -config /tmp/min-openssl-config.cnf
$ cat cert.pem key.pem > server.pem                                                                             
$ cp /tmp/cert.pem /etc/pki/ca-trust/source/anchors/cert.pem
$ update-ca-trust
```

```
$ cat /tmp/min-openssl-config.cnf
# Minimal openssl configuration file containing Subject Alternative Name (SAN)  
                                                                                
[ req ]                                                                         
distinguished_name = dn                                                         
                                                                                
[ dn ]                                                                          
basicConstraints = CA:FALSE                                                     
keyUsage = nonRepudiation, digitalSignature, keyEncipherment                    
subjectAltName = @alt_names                                                     
                                                                                
[alt_names]                                                                     
DNS.1 = localhost
```

removing `/etc/pki/ca-trust/source/anchors/cert.pem` and running `update-ca-trust` does not seem to help with this issue. Probably `update-ca-trust` is updating some other places as well that it fails to restore. Or maybe `curl` is caching something? :thinking:

**Comment by martinpitt** [comment]
[testCreate failure](https://logs.cockpit-project.org/logs/pull-13699-20200313-113500-a3ac6454-debian-testing/log.html#247), on [fedora-31](https://logs.cockpit-project.org/logs/pull-13699-20200313-113500-a3ac6454-fedora-31/log.html#247) as well. But it succeeded on fedora-31/firefox, and on RHEL, so it could just be another one of these notorious flakes. Retrying.

File: test/verify/machineslib.py [hunk 3]
```diff
@@ -201,6 +201,15 @@ def tearDown(self):
         # which need to be adjusted before we can do that.
         pass
 
+    def addIface(self, name):
+        self.machine.execute(r"""
+            mkdir -p /run/udev/rules.d/ &&
+            echo 'ENV{ID_NET_DRIVER}=="veth", ENV{INTERFACE}=="%s", ENV{NM_UNMANAGED}="0"' > /run/udev/rules.d/99-nm-veth-test.rules &&
+            udevadm control --reload &&
+            ip link add name %s type veth
+            """ % (name, name))
```
Comment (martinpitt) [inline test/verify/machineslib.py]: This is a bit ugly. With `%` you can use `"%(name)s" % { "name": name }`, or use `.format() with `{0}` twice.

File: test/verify/machineslib.py [hunk 4]
```diff
@@ -1479,7 +1489,8 @@ def installWithErrorTest(dialog):
         # Fake the osinfo-db data in order that it will allow spawn the installation - of course we don't expect it to succeed -
         # we just need to check that the VM was spawned
         fedora_28_xml = self.machine.execute("cat /usr/share/osinfo/os/fedoraproject.org/fedora-28.xml")
-        self.machine.execute("cat /usr/share/osinfo/os/fedoraproject.org/fedora-28.xml > /tmp/fedora-28.xml")
+        self.machine.execute("cp /usr/share/osinfo/os/fedoraproject.org/fedora-28.xml /tmp/fedora-28.xml")
+        self.addCleanup(self.machine.execute, "mv /tmp/fedora-28.xml /usr/share/osinfo/os/fedoraproject.org/fedora-28.xml")
```
Comment (martinpitt) [inline test/verify/machineslib.py]: Overwriting /usr is brittle and error -prone, and you couldn't do that on fedora-coreos. Seems better to get into the habit of bind-mounting it instead of `cp`ing.

File: test/verify/machineslib.py [hunk 5]
```diff
@@ -1708,6 +1719,7 @@ def installWithErrorTest(dialog):
             "virsh vol-create-as poolDisk sda1 1024"
         ]
         self.machine.execute(" && ".join(cmds))
+        self.addCleanup(self.machine.execute, "rm -f /tmp/poolDiskImages")
```
Comment (martinpitt) [inline test/verify/machineslib.py]: Could this rather output to the temp storage dir?

File: test/verify/machineslib.py [hunk 6]
```diff
@@ -1809,6 +1823,7 @@ def installWithErrorTest(dialog):
 
             # Disable selinux because it makes TFTP directory inaccesible and we don't want sophisticated configuration for tests
             self.machine.execute("if type selinuxenabled >/dev/null 2>&1 && selinuxenabled; then setenforce 0; fi")
+            self.addCleanup(self.machine.execute, "setenforce 1")
```
Comment (martinpitt) [inline test/verify/machineslib.py]: Ugh, how ugly -- let's fix that properly with setting the correct type with `chcon`. But let's do that in a separate PR, I made a note in #13717

File: test/verify/machineslib.py [hunk 7]
```diff
@@ -201,6 +201,15 @@ def tearDown(self):
         # which need to be adjusted before we can do that.
         pass
 
+    def addIface(self, name):
```
Comment (martinpitt) [inline test/verify/machineslib.py]: That's something which I'd rather like to have in testlib.py, so that the networking tests can use it as well. It should then also generate a name instead of asking the caller to make up one. I. e. it should keep a global counter or a set of used names or similar. But I'm happy to do this as a follow-up. Added to #13717

File: test/verify/machineslib.py [hunk 8]
```diff
@@ -1905,7 +1921,9 @@ def installWithErrorTest(dialog):
                                           {"Source": "Installation Source must not be empty"})
 
         # Test that removing virt-install executable will disable Create VM button
-        self.machine.execute('rm $(which virt-install)')
+        virt_install_bin = self.machine.execute("which virt-install").strip()
+        self.machine.execute('mv {0} {0}_backup'.format(virt_install_bin))
+        self.addCleanup(self.machine.execute, "mv {0}_backup {0}".format(virt_install_bin))
```
Comment (martinpitt) [inline test/verify/machineslib.py]: Again, let's please not mutilate /usr:
```
# mount -o bind /dev/null /usr/bin/virt-install
# which virt-install
/usr/bin/which: no virt-install in (/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/root/bin)
```

File: test/verify/machineslib.py (hunk 3 above)
Comment (marusak) [reply test/verify/machineslib.py]: `{0}` does not work as there are other `{...}` in the string. But using `{ "name": name }` is a good idea.

**Comment by marusak** [comment]
debian testing fails every time with:
```
internal error: process exited while connecting to monitor: 2020-03-15T16:02:22.119744Z qemu-system-x86_64: -drive file=https://localhost:8000/novell.iso,format=raw,if=none,id=drive-ide0-0-0,readonly=on: Unknown protocol 'https'
```

I'll try if it is reproducible also locally.

**Comment by marusak** [comment]
ah, so I was reading the logs wrongly. That "error" is displayed every time, even on success. The problem was that this test is flaky, the first run failed, then cleanup failed on `setenforce 1` and that broke the machine.

File: test/verify/machineslib.py [hunk 9]
```diff
@@ -1815,6 +1828,7 @@ def installWithErrorTest(dialog):
 
             # Disable selinux because it makes TFTP directory inaccesible and we don't want sophisticated configuration for tests
             self.machine.execute("if type selinuxenabled >/dev/null 2>&1 && selinuxenabled; then setenforce 0; fi")
+            self.addCleanup(self.machine.execute, "setenforce 1 || true")
```
Comment (martinpitt) [inline test/verify/machineslib.py]: That looks worrying. How exactly did it break? It's idempotent, i. e. if it's already in enforcing, the command still succeeds. So if re-enabling it fails, doesn't that mean the machine is already broken? Is there some way to reproduce that scenario? I have a feeling we need to clean up harder here.

Also, I want to get rid of this completely, see issue #13717. I think I should get to that today.

Comment (marusak) [reply test/verify/machineslib.py]:
```
root@debian:~# setenforce 1
setenforce: SELinux is disabled
root@debian:~# echo $?
1
```
It should be wrapped in `if type selinuxenabled >/dev/null 2>&1 && selinuxenabled; then setenforce 1; fi` as we disable selinux, but this was a quick workaround as there is #13717 to resolve this properly anyway.

Comment (martinpitt) [reply test/verify/machineslib.py]: Running `testCreate` locally is a nuisance, but I manually did the TFTP server steps and left SELinux enabled. I originally expected that I have to do something like `chcon -t svirt_t` to the created directories, but I can create a working PXE boot VM just like that. Perhaps you can try to just  drop the setenforce 0 bit completely?

Comment (martinpitt) [reply test/verify/machineslib.py]: But doing what you said is ok as well. We can then try and drop it completely in a separate PR, if you prefer.

File: test/verify/check-machines-dbus (hunk 2 above)
Comment (martinpitt) [reply test/verify/check-machines-dbus]: Part of  #13717, ok to ignore this for this PR.

**Comment by martinpitt** [comment]
[Some](https://logs.cockpit-project.org/logs/pull-13699-20200316-081649-3d9aa12f-fedora-31/log.html#247) [testCreate flakes](https://logs.cockpit-project.org/logs/pull-13699-20200316-081649-3d9aa12f-rhel-8-2/log.html#247), similar to the ones I get locally. But the SELinux bit does seem to work. :tada:
